checkout-core/src/main/java/com/adyen/checkout/core/api/Connection.kt

```diff
@@ -0,0 +1,157 @@
+ *
+ * Created by caiof on 17/12/2020.
+ */
+package com.adyen.checkout.core.api
+
+import com.adyen.checkout.core.BuildConfig
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InputStream
+import java.net.HttpURLConnection
+import java.nio.charset.StandardCharsets
+import java.util.*
+import java.util.concurrent.Callable
+
+/**
+ * A wrapper for a callable network connection.
+ *
+ * @param <T> The type of the connection return.
+ * @param url The URl used to make this Connection.
+ */
+abstract class Connection<T> protected constructor(
+    protected val url: String
+) : Callable<T> {
+
+    private var urlConnection: HttpURLConnection? = null
+
+    /**
+     * Performs an URL connection using HTTP GET.
+     *
+     * @param headers The headers of the connection.
+     * @return The byte array of the response
+     * @throws IOException In case an IO error happens.
+     */
+    @Suppress("TooGenericExceptionThrown")
+    @Throws(IOException::class)
+    protected operator fun get(headers: Map<String, String> = emptyMap()): ByteArray {
+        if (urlConnection != null) {
+            throw RuntimeException("Connection already initiated")
+        }
+        return try {
+            val connection = getUrlConnection(url, headers, HttpMethod.GET)
+            urlConnection = connection
+            connection.connect()
+            handleResponse(connection)
+        } finally {
+            urlConnection?.disconnect()
+        }
+    }
+
+    /**
+     * Performs an URL connection using HTTP POST.
+     *
+     * @return The byte array of the response
+     * @throws IOException In case an IO error happens.
+     */
+    @Suppress("TooGenericExceptionThrown")
+    @Throws(IOException::class)
+    protected fun post(headers: Map<String, String>, data: ByteArray): ByteArray {
+        if (urlConnection != null) {
+            throw RuntimeException("Connection already initiated")
+        }
+        return try {
+            val connection = getUrlConnection(url, headers, HttpMethod.POST)
+            urlConnection = connection
+            connection.connect()
+            connection.outputStream.use { outputStream ->
+                outputStream.write(data)
+                outputStream.flush()
+            }
+            handleResponse(connection)
+        } finally {
+            urlConnection?.disconnect()
+        }
+    }
+
+    private fun parseException(errorBytes: ByteArray?): IOException {
+        var message: String? = null
+        if (errorBytes != null) {
+            message = errorBytes.toString(CHARSET)
+        }
+        return IOException(message)
+    }
+
+    @Throws(IOException::class)
+    private fun getUrlConnection(url: String, headers: Map<String, String>, httpMethod: HttpMethod): HttpURLConnection {
+        val urlConnection = HttpUrlConnectionFactory.INSTANCE.createHttpUrlConnection(url)
```

Comment:
Can be replaced with `return HttpUrlConnectionFactory.INSTANCE.createHttpUrlConnection(url).apply{ ... } `